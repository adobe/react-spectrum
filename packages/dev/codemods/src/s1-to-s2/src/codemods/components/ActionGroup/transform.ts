import {addComponentImport} from '../../shared/utils';
import {commentOutProp} from '../../shared/transforms';
import {getComponents} from '../../../getComponents';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

let availableComponents = getComponents();

/**
 * Transforms ActionGroup:
 * - Comment out overflowMode (it has not been implemented yet).
 * - Comment out buttonLabelBehavior (it has not been implemented yet).
 * - Comment out summaryIcon (it has not been implemented yet).
 * - Use ActionButtonGroup if no selection.
 * - Use ToggleButtonGroup if selection is used.
 * - Update root level onAction to onPress on each ActionButton.
 * - Apply isDisabled directly on each ActionButton/ToggleButton instead of disabledKeys.
 * - Convert dynamic collections render function to items.map.
 */
export default function transformActionGroup(path: NodePath<t.JSXElement>) {
  // Comment out overflowMode
  commentOutProp(path, {propName: 'overflowMode'});

  // Comment out buttonLabelBehavior
  commentOutProp(path, {propName: 'buttonLabelBehavior'});

  // Comment out summaryIcon
  commentOutProp(path, {propName: 'summaryIcon'});

  let selectionModePath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'selectionMode') as NodePath<t.JSXAttribute> | undefined;
  let selectionMode = t.isStringLiteral(selectionModePath?.node.value) ? selectionModePath.node.value.value : 'none';
  let newComponentName, childComponentName;
  if (selectionMode === 'none') {
    // Use ActionButtonGroup if no selection
    newComponentName = 'ActionButtonGroup';
    childComponentName = 'ActionButton';
    selectionModePath?.remove();
  } else {
    // Use ToggleButtonGroup if selection is used
    newComponentName = 'ToggleButtonGroup';
    childComponentName = 'ToggleButton';
  }

  let localName = newComponentName;
  if (availableComponents.has(newComponentName)) {
    let program = path.findParent((p) => t.isProgram(p.node)) as NodePath<t.Program>;
    localName = addComponentImport(program, newComponentName);
  }

  let localChildName = childComponentName;
  if (availableComponents.has(childComponentName)) {
    let program = path.findParent((p) => t.isProgram(p.node)) as NodePath<t.Program>;
    localChildName = addComponentImport(program, childComponentName);
  }


  // Convert dynamic collection to an array.map.
  let items = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'items') as NodePath<t.JSXAttribute> | undefined;
  let itemArg: t.Identifier | undefined;
  if (items && t.isJSXExpressionContainer(items.node.value) && t.isExpression(items.node.value.expression)) {
    let child = path.get('children').find(c => c.isJSXExpressionContainer());
    if (child && child.isJSXExpressionContainer() && t.isFunction(child.node.expression)) {
      let arg = child.node.expression.params[0];
      if (t.isIdentifier(arg)) {
        itemArg = arg;
      }

      child.replaceWith(
        t.jsxExpressionContainer(
          t.callExpression(
            t.memberExpression(
              items.node.value.expression,
              t.identifier('map')
            ),
            [child.node.expression]
          )
        )
      );
    }
  }
  items?.remove();

  let onAction = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'onAction') as NodePath<t.JSXAttribute> | undefined;

  // Pull disabledKeys prop out into a variable, converted to a Set.
  // Then we can check it in the isDisabled prop of each item.
  let disabledKeysPath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'disabledKeys') as NodePath<t.JSXAttribute> | undefined;
  let disabledKeys: t.Identifier | undefined;
  if (disabledKeysPath && t.isJSXExpressionContainer(disabledKeysPath.node.value) && t.isExpression(disabledKeysPath.node.value.expression)) {
    disabledKeys = path.scope.generateUidIdentifier('disabledKeys');
    path.scope.push({
      id: disabledKeys,
      init: t.newExpression(t.identifier('Set'), [disabledKeysPath.node.value.expression]),
      kind: 'let'
    });
    disabledKeysPath.remove();
  }

  path.traverse({
    JSXElement(child) {
      if (t.isJSXIdentifier(child.node.openingElement.name) && child.node.openingElement.name.name === 'Item') {
        // Replace Item with ActionButton or ToggleButton.
        let childNode = t.cloneNode(child.node);
        childNode.openingElement.name = t.jsxIdentifier(localChildName);
        if (childNode.closingElement) {
          childNode.closingElement.name = t.jsxIdentifier(localChildName);
        }

        // If there is no key prop and we are using dynamic collections, add a default computed from item.key ?? item.id.
        let key = childNode.openingElement.attributes.find(attr => t.isJSXAttribute(attr) && attr.name.name === 'key') as t.JSXAttribute | undefined;
        if (!key && itemArg) {
          let id = t.jsxExpressionContainer(
            t.logicalExpression(
              '??',
              t.memberExpression(itemArg, t.identifier('key')),
              t.memberExpression(itemArg, t.identifier('id'))
            )
          );

          key = t.jsxAttribute(
            t.jsxIdentifier('key'),
            id
          );

          childNode.openingElement.attributes.push(key);
        }

        // If this is a ToggleButtonGroup, add an id prop in addition to key when needed.
        if (key && newComponentName === 'ToggleButtonGroup') {
          // If we are in an array.map we need both key and id. Otherwise, we only need id.
          if (itemArg) {
            childNode.openingElement.attributes.push(t.jsxAttribute(t.jsxIdentifier('id'), key.value));
          } else {
            key.name.name = 'id';
          }
        }

        let keyValue: t.Expression | undefined = undefined;
        if (key && t.isJSXExpressionContainer(key.value) && t.isExpression(key.value.expression)) {
          keyValue = key.value.expression;
        } else if (key && t.isStringLiteral(key.value)) {
          keyValue = key.value;
        }

        // Add an onPress to each item that calls the previous onAction, passing in the key.
        if (onAction && t.isJSXExpressionContainer(onAction.node.value) && t.isExpression(onAction.node.value.expression)) {
          childNode.openingElement.attributes.push(
            t.jsxAttribute(
              t.jsxIdentifier('onPress'),
              t.jsxExpressionContainer(
                keyValue 
                  ? t.arrowFunctionExpression([], t.callExpression(onAction.node.value.expression, [keyValue]))
                  : onAction.node.value.expression
              )
            )
          );
        }

        // Add an isDisabled prop to each item, testing whether it is in disabledKeys.
        if (disabledKeys && keyValue) {
          childNode.openingElement.attributes.push(
            t.jsxAttribute(
              t.jsxIdentifier('isDisabled'),
              t.jsxExpressionContainer(
                t.callExpression(
                  t.memberExpression(
                    disabledKeys,
                    t.identifier('has')
                  ),
                  [keyValue]
                )
              )
            )
          );
        }

        child.replaceWith(childNode);
      }
    }
  });

  onAction?.remove();

  path.node.openingElement.name = t.jsxIdentifier(localName);
  if (path.node.closingElement) {
    path.node.closingElement.name = t.jsxIdentifier(localName);
  }
} 
