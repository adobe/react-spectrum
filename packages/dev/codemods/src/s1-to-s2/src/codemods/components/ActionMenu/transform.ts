import {commentOutProp} from '../../shared/transforms';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

/**
 * Transforms ActionMenu:
 * - Comment out closeOnSelect (it has not been implemented yet).
 * - Comment out trigger (it has not been implemented yet).
 */
export default function transformActionMenu(path: NodePath<t.JSXElement>) {
  // Comment out closeOnSelect
  commentOutProp(path, {propName: 'closeOnSelect'});

  // Comment out trigger
  commentOutProp(path, {propName: 'trigger'});
} 
