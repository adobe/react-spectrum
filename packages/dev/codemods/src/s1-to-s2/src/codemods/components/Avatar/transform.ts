import {commentOutProp} from '../../shared/transforms';
import {getName} from '../../shared/utils';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

function updateAvatarSize(
  path: NodePath<t.JSXElement>
) {
  if (
    t.isJSXElement(path.node) &&
    t.isJSXIdentifier(path.node.openingElement.name) &&
    getName(path, path.node.openingElement.name) === 'Avatar'
  ) {
    let sizeAttrPath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'size') as NodePath<t.JSXAttribute>;
    if (sizeAttrPath) {
      let value = sizeAttrPath.node.value;
      if (value?.type === 'StringLiteral') {
        const avatarDimensions = {
          'avatar-size-50': 16,
          'avatar-size-75': 18,
          'avatar-size-100': 20,
          'avatar-size-200': 22,
          'avatar-size-300': 26,
          'avatar-size-400': 28,
          'avatar-size-500': 32,
          'avatar-size-600': 36,
          'avatar-size-700': 40
        };
        let val = avatarDimensions[value.value as keyof typeof avatarDimensions];
        if (val != null) {
          sizeAttrPath.node.value = t.jsxExpressionContainer(t.numericLiteral(val));
        }
      }
    }
  }
}

/**
 * Transforms Avatar:
 * - Comment out isDisabled (it has not been implemented yet).
 * - Update size to be a pixel value if it currently matches 'avatar-size-*'.
 */
export default function transformAvatar(path: NodePath<t.JSXElement>) {
  // Comment out isDisabled
  commentOutProp(path, {propName: 'isDisabled'});

  // Update size to be a pixel value if it currently matches 'avatar-size-*'
  updateAvatarSize(path);
} 
