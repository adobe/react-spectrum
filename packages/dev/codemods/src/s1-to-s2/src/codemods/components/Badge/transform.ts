import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';
import {updatePropNameAndValue} from '../../shared/transforms';

/**
 * Transforms Badge:
 * - Change variant="info" to variant="informative".
 */
export default function transformBadge(path: NodePath<t.JSXElement>) {
  // Change variant="info" to variant="informative"
  updatePropNameAndValue(path, {
    oldPropName: 'variant',
    oldPropValue: 'info',
    newPropName: 'variant',
    newPropValue: 'informative'
  });
} 
