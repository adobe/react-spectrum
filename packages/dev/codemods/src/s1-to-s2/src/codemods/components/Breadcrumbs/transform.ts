import {
  addCommentToElement,
  commentOutProp,
  removeProp
} from '../../shared/transforms';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

/**
 * Transforms Breadcrumbs:
 * - Comment out showRoot (it has not been implemented yet).
 * - Comment out isMultiline (it has not been implemented yet).
 * - Comment out autoFocusCurrent (it has not been implemented yet).
 * - Remove size="S" (Small is no longer a supported size in Spectrum 2).
 * - Add comment to wrap in nav element if needed.
 */
export default function transformBreadcrumbs(path: NodePath<t.JSXElement>) {
  // Comment out showRoot
  commentOutProp(path, {propName: 'showRoot'});

  // Comment out isMultiline
  commentOutProp(path, {propName: 'isMultiline'});

  // Comment out autoFocusCurrent
  commentOutProp(path, {propName: 'autoFocusCurrent'});

  // Remove size="S"
  removeProp(path, {propName: 'size', propValue: 'S'});

  // Add comment to wrap in nav element if needed
  addCommentToElement(path, {
    comment: 'S2 Breadcrumbs no longer includes a nav element by default. You can wrap the Breadcrumbs component in a nav element if needed.'
  });
} 
