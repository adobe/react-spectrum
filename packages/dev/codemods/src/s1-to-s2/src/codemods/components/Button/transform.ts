import {NodePath} from '@babel/traverse';
import {
  removeProp,
  updateComponentIfPropPresent,
  updatePropName,
  updatePropNameAndValue,
  updatePropValueAndAddNewPropName
} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms Button:
 * - Change variant="cta" to variant="accent"
 * - Change variant="overBackground" to variant="primary" staticColor="white"
 * - Change style to fillStyle
 * - Remove isQuiet (it is no longer supported in Spectrum 2)
 * - If href is present, the Button should be converted to a LinkButton
 * - Remove elementType (it is no longer supported in Spectrum 2).
 */
export default function transformButton(path: NodePath<t.JSXElement>) {
  // Change variant="cta" to variant="accent"
  updatePropNameAndValue(path, {
    oldPropName: 'variant',
    oldPropValue: 'cta',
    newPropName: 'variant',
    newPropValue: 'accent'
  });

  // Change variant="overBackground" to variant="primary" staticColor="white"
  updatePropValueAndAddNewPropName(path, {
    oldPropName: 'variant',
    oldPropValue: 'overBackground',
    newPropName: 'variant',
    newPropValue: 'primary',
    additionalPropName: 'staticColor',
    additionalPropValue: 'white'
  });

  // Change style to fillStyle
  updatePropName(path, {
    oldPropName: 'style',
    newPropName: 'fillStyle'
  });

  // Remove isQuiet
  removeProp(path, {propName: 'isQuiet'});

  // If href is present, the Button should be converted to a LinkButton
  updateComponentIfPropPresent(path, {
    propName: 'href',
    newComponentName: 'LinkButton'
  });

  // Remove elementType
  removeProp(path, {propName: 'elementType'});
}
