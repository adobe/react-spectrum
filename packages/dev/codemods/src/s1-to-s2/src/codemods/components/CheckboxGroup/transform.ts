import {NodePath} from '@babel/traverse';
import {removeProp} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms CheckboxGroup:
 * - Remove showErrorIcon (it has been removed due to accessibility issues).
 */
export default function transformCheckboxGroup(path: NodePath<t.JSXElement>) {
  // Remove showErrorIcon
  removeProp(path, {propName: 'showErrorIcon'});
} 
