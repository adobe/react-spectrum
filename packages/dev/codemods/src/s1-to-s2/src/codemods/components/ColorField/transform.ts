import {NodePath} from '@babel/traverse';
import {removeProp, updatePropNameAndValue} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms ColorField:
 * - Remove isQuiet (it is no longer supported in Spectrum 2).
 * - Remove placeholder (it has been removed due to accessibility issues).
 * - Change validationState="invalid" to isInvalid.
 * - Remove validationState="valid" (it is no longer supported in Spectrum 2).
 */
export default function transformColorField(path: NodePath<t.JSXElement>) {
  // Remove isQuiet
  removeProp(path, {propName: 'isQuiet'});

  // Remove placeholder
  removeProp(path, {propName: 'placeholder'});

  // Change validationState="invalid" to isInvalid
  updatePropNameAndValue(path, {
    oldPropName: 'validationState',
    oldPropValue: 'invalid',
    newPropName: 'isInvalid',
    newPropValue: true
  });

  // Remove validationState="valid"
  removeProp(path, {propName: 'validationState', propValue: 'valid'});
} 
