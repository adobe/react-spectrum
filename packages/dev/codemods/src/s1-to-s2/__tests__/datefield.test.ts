// @ts-ignore
import {defineSnapshotTest} from 'jscodeshift/dist/testUtils';
import transform from '../src/codemods/codemod';

const test = (name: string, input: string) => {
  defineSnapshotTest(transform, {}, input, name);
};

test('Removes isQuiet', `

import {DateField} from '@adobe/react-spectrum';
let isQuiet = true;
let props = {isQuiet: true};
<div>
  <DateField isQuiet />
  <DateField isQuiet={true} />
  <DateField isQuiet={false} />
  <DateField isQuiet={isQuiet} />
  <DateField isQuiet={'foo' === 'foo'} />
  <DateField {...props} />
</div>
`);


test('changes validationState to isInvalid or nothing', `
  import {DateField} from '@adobe/react-spectrum';
  let validationState = 'invalid';
  let props = {validationState: 'invalid'};
  <div>
    <DateField validationState="invalid" />
    <DateField validationState="valid" />
    <DateField validationState={validationState} />
    <DateField {...props} />
  </div>
`);
