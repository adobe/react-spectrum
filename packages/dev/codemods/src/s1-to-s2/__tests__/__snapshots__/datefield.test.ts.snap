// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does nothing 1`] = `
"import { DateField } from "@react-spectrum/s2";

<div>
  <DateField label="Event date" />
</div>"
`;
